File: Binary_tree_level_traversal.py
# DFS- approach- using stack- recursion

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        result = []
        depth = 0
        self.dfs(root,result,depth)
        return result
    
    def dfs(self, root, result, depth):
        if root == None:
            return None
        
        if len(result) <= depth:
            result.append([])
        result[depth].append(root.val)
        
        self.dfs(root.left, result, depth+1)
        self.dfs(root.right, result, depth+1)
        
Time complexity- O(N) + O(H)- stack memory      
Space complexity- O(H) for recursion and O(BReadth) for iterartive



# BFS - using queue- iterative 

class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        
        result = []
        q = deque([root])
        
        if root == None:
            return None
        
        
        while q:
            
            level = []
            
            for i in range(0, len(q)):
                
                
                temp = q.popleft()
                
                if temp.left != None:
                    q.append(temp.left)
                    
                if temp.right != None:
                    q.append(temp.right)
                
                level.append(temp.val)
            
            r.append(level[-1])
            result.append(level)
            
        
        return result

File: BinaryTreeLevelTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
 //Time complexity is O(N)
 //Space complexity is O(N)
class Solution {
    List<List<Integer>> result = new ArrayList();
    public List<List<Integer>> levelOrder(TreeNode root) {
        if(root==null){
            return result;
        }
        helper(root, 1);
        return result;
    }
    public void helper(TreeNode root, int depth){
        if(root==null){
            return;
        }
        if(result.size()<depth){
            result.add(new ArrayList());
        }
        result.get(depth-1).add(root.val);
        helper(root.left, depth+1);
        helper(root.right, depth+1);
    }
}